### PR TITLE
adc: Convert mcux adc16 driver to the new adc api

### PR DESCRIPTION
**Pull request by MaureenHelm**
Converts the mcux adc16 driver to the new adc api and updates tests accordingly.

Depends on #7691

### CODE REVIEW

**Comment by anangl** [comment]
@MaureenHelm My original intention was to make channels within API independent of any physical ADC inputs (now I know it was not necessarily a good idea). That's why in the test I used fixed values for channel identifiers (0 and 2 because I just wanted to check that non-continuous channel identifiers are also handled correctly).
I updated my PR, so after rebasing, your second commit should be no longer needed. You'll need to adjust a bit the third one. I hope it won't be a big problem.

File: drivers/adc/adc_mcux_adc16.c [hunk 1]
```diff
@@ -1,69 +1,171 @@
...
+{
+	const struct mcux_adc16_config *config = dev->config->config_info;
+	struct mcux_adc16_data *data = dev->driver_data;
+	adc16_hardware_average_mode_t mode;
+	int error;
+
+	if (sequence->resolution != 12) {
+		SYS_LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
```
Comment (anangl) [inline drivers/adc/adc_mcux_adc16.c]: You should also check if any channels were selected and if only the valid ones.

File: drivers/adc/adc_mcux_adc16.c [hunk 2]
```diff
@@ -1,69 +1,171 @@
...
-
-		data->channel_group = channel_group;
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct mcux_adc16_data *data =
+		CONTAINER_OF(ctx, struct mcux_adc16_data, ctx);
 
-		k_sem_take(&data->sync, K_FOREVER);
+	if (ctx->sampling_index == 0) {
+		data->buffer = ctx->sequence->buffer;
```
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply drivers/adc/adc_mcux_adc16.c]: Good catch, thank you. The improved test did make it fail, and now it passes after fixing up the buffer pointer initialization/reinitialization.

**Comment by MaureenHelm** [comment]
> I updated my PR, so after rebasing, your second commit should be no longer needed. You'll need to adjust a bit the third one. I hope it won't be a big problem.

Not a problem. I've rebased and removed the now-obsolete commit.

**Pull request closed without merging**